### ISSUE TITLE
Server/client architecture for copy and move not working

### ISSUE THREAD

**Issue by og900aero** [gokcehan/lf#1037]
I use lf with custom copy and move command: programmed copy and move wit hpaste aka Vifm, so one key to select for copy and move (space), and fo different key for copy-paste (p) and copy-move (P):

```
cmd paste %{{
    set -- $(cat ~/.local/share/lf/files)
    mode="$1"
    shift
    case "$mode" in
        copy)
            # rsync -ah --info=progress2 -- "$@" . |
            gcp -rfp -- "$@" . |
            stdbuf -i0 -o0 -e0 tr '\r' '\n' |
            while IFS= read -r line; do
                lf -remote "send $id echo $line"
            done
            ;;
        move) mv -n -- "$@" .;;
    esac
    rm ~/.local/share/lf/files
    cd $PWD
    lf -remote "send clear"
    lf -remote "send load"
}}
map <space> :toggle; calcdirsize; down
map p :copy; paste
map P :cut; paste
```

This working great with one instance lf, but doesnt work two instance. So I open another lf. Use space to select files or folders, then I move to the second lf (I use bspwm tiling windows manager), és press p, and doesnt happend. I feel that this server-client solution only works properly with the built-in commands, and if someone modifies it, it will be a problem. Couldn't the tab solution be better than what ranger, nnn, joshuto, etc. do you use it too?
I really like lf, because it can be customized in detail, but this server-client setup is disastrous for larger copies, if it works at all, not to mention that you have to open another window, which takes up unnecessary space, in the case of tmux, it reduces the visible area.

**Comment by DusanLesan** [comment]
I have never tried to modify copy paste, but I think that selection buffers are not shared between instances. So when you copy, there is nothing in current selection buffer. You could maybe try to use remote commands to get other instance to move selected items into move or copy buffer. In that case you would need to specify id of other instance since commanding copy on all instances would probably not work ok. 
Something like this:
```
For each other instance of lf
  call copy/move,
  get all items in files buffer list (maybe also clear it) and keep it in new list
  move to next instance
paste using a new list
```
I hope that helps.
Also, could you explain why do you have to open another window for copying. Majority of times I use only a single client.

**Comment by og900aero** [comment]
_get all items in files buffer list (maybe also clear it) and keep it in new list
  move to next instance_
  
  I don't know how to do this in the case of lf.
  
_Also, could you explain why do you have to open another window for copying. Majority of times I use only a single client._

It is necessary to open another window because there are cases when it is much more convenient to navigate to the destination in another window, on another tab, because it is advisable to stay at the source location because there are other things to do. In such cases, it is much more time-consuming and tedious to constantly walk from one directory to another, as if there were a more expedient and simpler solution than that followed by other file management programs. See vifm, ranger, joshuto, mc, nnn, etc

**Comment by DusanLesan** [comment]
> I don't know how to do this in the case of lf.

You would need to get pids of all lf instances and loop through them. Something like this:
```
cmd paste ${{
	# Get all pids
	pids="$(pgrep -f "lf -last-dir-path")"
	# Loop through the pids
	for id in $pids; do
		# Send command to all instances to save currently selected items
		lf -remote "send $id echoSelected"
	done
	# Filter out duplicates
	# Do paste stuff
}}

cmd echoSelected ${{
	# If there are selected items, add them to the file. Each instance should get this command
	if [ -n "$fs" ]; then
		for path in "$fs"; do
			echo "$path" >> ~/.local/share/lf/files
		done
	fi
}}
```

> See vifm, ranger, joshuto, mc, nnn, etc

I have not used this apps in ages so I cannot remember how it is handled in them. Have you tried marks? You can save current location with mark-save and then load that with mark-load

**Comment by og900aero** [comment]
> for id in $pids; do
> 		# Send command to all instances to save currently selected items
> 		lf -remote "send $id echoSelected"
> 	done

I try this method, but nothing happend.

**Comment by DusanLesan** [comment]
> I try this method, but nothing happend.

I wrote that as example but it should work like that. That should take all selected items and add them to ~/.local/share/lf/files. After calling it on every instance, they should append it to that file and later you can use it for doing your custom paste logic. Try to rename that function so you are not replacing the built in but making your own. Also be sure that at least one instance have selected items

**Comment by og900aero** [comment]
> > I try this method, but nothing happend.
> 
> I wrote that as example but it should work like that. That should take all selected items and add them to ~/.local/share/lf/files. After calling it on every instance, they should append it to that file and later you can use it for doing your custom paste logic. Try to rename that function so you are not replacing the built in but making your own. Also be sure that at least one instance have selected items

As I look at the contents of ~/.local/share/lf/files, it won't even work in the setup I use (I select the necessary files and folders, then when pasting I decide whether to copy or move, i.e. just the other way around, as the original setup), because the very first entry of the file contents contains whether it is copying or moving, and only then does it list the files it will be executed on, i.e. just the other way around, as I would use it. So the server-client solution seems like a dead end to me.

**Comment by DusanLesan** [comment]
> As I look at the contents of ~/.local/share/lf/files, it won't even work in the setup I use (I select the necessary files and folders, then when pasting I decide whether to copy or move, i.e. just the other way around, as the original setup), because the very first entry of the file contents contains whether it is copying or moving, and only then does it list the files it will be executed on, i.e. just the other way around, as I would use it. So the server-client solution seems like a dead end to me.

Ok. I do not understand the issue. You are able to collect list of all selected files in any file of your choosing from other instances and you are able to process that as you see fit in a active instance when you call custom paste command. That sounds as enough info to me.

**Comment by og900aero** [comment]
The basic problem is that I am not using the copy and move command as the lf developer intended:
1. I select what I want to copy/move with space
2. Press y or d to select whether to copy or move
3. I go to a specified folder and press a p to paste

I use it like this instead (it's nicer and more reasonable):
1. I select what I want to copy/move with space
2. I go to a given folder and press a p for copy and paste, P ha for move and paste.

I assigned the following command to this p and P:
```
map p :copy; paste
map P :cut; paste

```
The paste command copied from the wiki (asynchronous copying and using gcp instead of cp):
```
cmd paste %{{
    set -- $(cat ~/.local/share/lf/files)
    mode="$1"
    shift
    case "$mode" in
        copy)
            gcp -rfp -- "$@" . |
            stdbuf -i0 -o0 -e0 tr '\r' '\n' |
            while IFS= read -r line; do
                lf -remote "send $id echo $line"
            done
            ;;
        move) mv -n -- "$@" .;;
    esac
    rm ~/.local/share/lf/files
    cd $PWD
    lf -remote "send clear"
    lf -remote "send load"
}}
```
In this configuration, the server-client configuration is unfortunately inoperable, or the egas command would have to be reprogrammed, which is beyond my capabilities (and I have already programmed many things in it, practically everything that I used in vifm and ranger.)

**Comment by DusanLesan** [comment]
Maybe this could help you do it:
```
cmd signalMove ${{
	:> ~/.local/share/lf/custom_file_list
	pids="$(pgrep -f "^lf -last-dir-path")"
	for pid in $pids; do
		lf -remote "send $pid saveSelected"
	done
	lf -remote "send $id moveFromList"
}}

cmd moveFromList ${{
	for file in $(< ~/.local/share/lf/custom_file_list); do
		mv $file .
	done
	lf -remote "send clear"
	lf -remote "send load"
}}

cmd signalCopy ${{
	:> ~/.local/share/lf/custom_file_list
	pids="$(pgrep -f "^lf -last-dir-path")"
	for pid in $pids; do
		lf -remote "send $pid saveSelected"
	done
	lf -remote "send $id copyFromList"
}}

cmd copyFromList ${{
	for file in $(< ~/.local/share/lf/custom_file_list); do
		cp $file .
	done
	lf -remote "send clear"
	lf -remote "send load"
}}

cmd saveSelected ${{
	if [ -n "$fs" ]; then
		for path in "$fs"; do
			echo "$path" >> ~/.local/share/lf/custom_file_list
		done
	fi
}}

map P :signalMove
map p :signalCopy
```
There is a lot of duplicate code so it probably can we rewritten better. It does not handle directories when copying, duplicates in list are not filtered out but you can do it if the base idea is matching your requirements

**Comment by og900aero** [comment]
> Maybe this could help you do it:
> 
> ```
> cmd signalMoveOrCopy ${{
> 	:> ~/.local/share/lf/custom_file_list
> 	pids="$(pgrep -f "^lf -last-dir-path")"
> 	for pid in $pids; do
> 		lf -remote "send $pid saveSelected"
> 	done
> 	lf -remote "send $id processList $1"
> }}
> 
> cmd processList ${{
> 	for file in $(< ~/.local/share/lf/custom_file_list); do
> 		if [[ $1 == "copy" ]]; then
> 			cp $file .
> 		else
> 			mv $file .
> 		fi
> 	done
> 	lf -remote "send clear"
> 	lf -remote "send load"
> }}
> 
> cmd saveSelected ${{
> 	if [ -n "$fs" ]; then
> 		for path in "$fs"; do
> 			echo "$path" >> ~/.local/share/lf/custom_file_list
> 		done
> 	fi
> }}
> 
> map P :signalMoveOrCopy move
> map p :signalMoveOrCopy copy
> ```
> 
> It does not handle directories when copying, duplicates in list are not filtered out but you can do it if the base idea is matching your requirements

Unfortunately, it doesn't work at all. I try only one instance at first, but get error message: running shell: exit status 1

**Comment by DusanLesan** [comment]
Is there an error message when you quit lf?

**Comment by DusanLesan** [comment]
>  So, the copy command should be rewritten only so that it not only pastes the selections found in the current window into the buffer, but also searches for the selections found in all running lf and pastes them into the buffer.

The code I wrote above does that. It processes all files from $fs variables from all instances.

**Comment by og900aero** [comment]
> > So, the copy command should be rewritten only so that it not only pastes the selections found in the current window into the buffer, but also searches for the selections found in all running lf and pastes them into the buffer.
> 
> The code I wrote above does that. It processes all files from $fs variables from all instances.

But doesnt work.

**Comment by DusanLesan** [comment]
`exit status 1` is not enough info to see why it is not working for you. Add some prints, see how far it gets to see what part is giving the error. Run parts piece by piece. For example:

- Does `pgrep -f "^lf -last-dir-path` return anything when instance is opened
- Is the pid valid
- Can you send it remote command
- Does `lf -remote "send $pid saveSelected"` save selected files as expected
- ...

**Comment by og900aero** [comment]
> * pgrep -f "^lf -last-dir-path

Opened two lf. andOpen a terminal window, and I issued the following command: pgrep -f "^lf -last-dir-path"
I get nothing. 
I run ps -aux | grep lf, and see the following:
```
shyciii     3399  0.1  0.6  57396 23796 ?        Ss   08:06   0:00 st -e lf -command cd /home/Data
shyciii     3409  0.0  0.2 708020  8532 pts/0    Ssl+ 08:06   0:00 lf -command cd /home/Data
shyciii     3428  0.0  0.1 706868  4248 ?        Ssl  08:06   0:00 lf -server
shyciii     3895  0.1  0.5  56160 22592 ?        Ss   08:09   0:00 st -e lf -command cd /home/Data
shyciii     3903  0.0  0.2 707444  8076 pts/1    Ssl+ 08:09   0:00 lf -command cd /home/Data
shyciii     3953  0.1  0.4  52732 19148 ?        Ss   08:09   0:00 st -e lf -command cd /home/Data
shyciii     3967  0.0  0.1 707700  7212 pts/2    Ssl+ 08:09   0:00 lf -command cd /home/Data
```

**Comment by DusanLesan** [comment]
Find any other way to get lf instance IDs.

**Comment by og900aero** [comment]
> Find any other way to get lf instance IDs than what I was using.

Okay, modified the pgrep. Now, the files to be copied are already entered in the custom_file_list file, but the copying itself does not work and there is no error message. So the following rows have something wrong:

```
cmd processList ${{
	for file in $(< ~/.local/share/lf/custom_file_list); do
		if [[ $1 == "copy" ]]; then
			cp $file .
		else
			mv $file .
		fi
	done
	lf -remote "send clear"
	lf -remote "send load"
}}
```

**Comment by og900aero** [comment]
> Does it work if you call `lf -remote "send <some id> processList copy`?

I rewrote your script a bit, and it works like this if only one file/directory needs to be copied:

cmd processList !{{
    load=$(cat ~/.local/share/lf/custom_file_list)
    if [ $1 = 'copy' ]; then
        cp -r "$load" .
    elif [ $1 = 'move' ]; then
        mv "$load" .
    fi
}}

But if I select more than one, an error message appears:
"cp: cannot stat '/home/Data/Audacity'$'\n''/home/Data/Benchmarks': No such file or directory"

Maybe a problem: $'\n'
It does not properly handle the fact that it is already a separate line, even though I entered in the lfrc: set ifs="\n"

**Comment by og900aero** [comment]
We have the solution and it works. I started from the idea of using the special insertion command from the wiki/tip (progress bar, although I rewrote it a bit), which works well, that is, the selection would have to be rewritten, so I modified the toggle command, and very simplet:

cmd mark_copy %{{
	newline='\n'
	echo "copy$newline$fs" > ~/.local/share/lf/files
	sed --quiet -i -e '/^$/d' -e 'p' ~/.local/share/lf/files
}}

map <space> :toggle; mark_copy; down

Then I modified the paste command a little, so that if I choose move, the copy command from files should be replaced by move. It works great.
Thank you for your help!
